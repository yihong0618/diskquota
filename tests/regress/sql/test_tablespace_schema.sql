-- Test schema
-- start_ignore
\! mkdir -p /tmp/schemaspc
-- end_ignore
CREATE SCHEMA spcs1;
DROP TABLESPACE  IF EXISTS schemaspc;
CREATE TABLESPACE schemaspc LOCATION '/tmp/schemaspc';
SET search_path TO spcs1;

CREATE TABLE a(i int) TABLESPACE schemaspc DISTRIBUTED BY (i);
INSERT INTO a SELECT generate_series(1,100);
-- expect insert fail
INSERT INTO a SELECT generate_series(1,100000);

SELECT diskquota.set_schema_tablespace_quota('spcs1', 'schemaspc','1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);
CREATE TABLE a2(i int) TABLESPACE schemaspc DISTRIBUTED BY (i);
-- expect insert fail
INSERT INTO a2 SELECT generate_series(1,100);

-- Test alter table set schema
CREATE SCHEMA spcs2;
ALTER TABLE spcs1.a SET SCHEMA spcs2;
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert succeed
INSERT INTO a2 SELECT generate_series(1,200);
-- expect insert succeed
INSERT INTO spcs2.a SELECT generate_series(1,200);
ALTER TABLE spcs2.a SET SCHEMA spcs1;
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert fail
INSERT INTO a SELECT generate_series(1,200);
SELECT schema_name, tablespace_name, quota_in_mb, nspsize_tablespace_in_bytes FROM diskquota.show_fast_schema_tablespace_quota_view WHERE schema_name = 'spcs1' and tablespace_name ='schemaspc';

-- Test alter tablespace
-- start_ignore
\! mkdir -p /tmp/schemaspc2
-- end_ignore
DROP TABLESPACE  IF EXISTS schemaspc2;
CREATE TABLESPACE schemaspc2 LOCATION '/tmp/schemaspc2';
ALTER TABLE a SET TABLESPACE schemaspc2;
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert succeed
INSERT INTO a SELECT generate_series(1,200);
ALTER TABLE a SET TABLESPACE schemaspc;
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert fail
INSERT INTO a SELECT generate_series(1,200);

-- Test update quota config
SELECT diskquota.set_schema_tablespace_quota('spcs1', 'schemaspc', '10 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert success
INSERT INTO a SELECT generate_series(1,100);
-- expect insert success
INSERT INTO a SELECT generate_series(1,1000000);
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);

-- Test delete quota config
SELECT diskquota.set_schema_tablespace_quota('spcs1', 'schemaspc', '-1 MB');
SELECT diskquota.wait_for_worker_new_epoch();
-- expect insert success
INSERT INTO a SELECT generate_series(1,100);

-- start_ignore
\! mkdir -p /tmp/schemaspc3
-- end_ignore
DROP TABLESPACE  IF EXISTS "Schemaspc3";
CREATE TABLESPACE "Schemaspc3" LOCATION '/tmp/schemaspc3';
CREATE SCHEMA "Spcs2";
SELECT diskquota.set_schema_tablespace_quota('"Spcs2"', '"Schemaspc3"', '-1 MB');

RESET search_path;
DROP TABLE spcs1.a2, spcs1.a;
DROP SCHEMA spcs1, spcs2;
DROP TABLESPACE schemaspc;
DROP TABLESPACE schemaspc2;
DROP TABLESPACE "Schemaspc3";
