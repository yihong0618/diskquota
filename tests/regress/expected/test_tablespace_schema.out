-- Test schema
-- start_ignore
\! mkdir -p /tmp/schemaspc
-- end_ignore
CREATE SCHEMA spcs1;
DROP TABLESPACE  IF EXISTS schemaspc;
NOTICE:  tablespace "schemaspc" does not exist, skipping
CREATE TABLESPACE schemaspc LOCATION '/tmp/schemaspc';
SET search_path TO spcs1;
CREATE TABLE a(i int) TABLESPACE schemaspc DISTRIBUTED BY (i);
INSERT INTO a SELECT generate_series(1,100);
-- expect insert fail
INSERT INTO a SELECT generate_series(1,100000);
SELECT diskquota.set_schema_tablespace_quota('spcs1', 'schemaspc','1 MB');
 set_schema_tablespace_quota 
-----------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);
ERROR:  tablespace: schemaspc, schema: spcs1 diskquota exceeded
CREATE TABLE a2(i int) TABLESPACE schemaspc DISTRIBUTED BY (i);
-- expect insert fail
INSERT INTO a2 SELECT generate_series(1,100);
ERROR:  tablespace: schemaspc, schema: spcs1 diskquota exceeded
-- Test alter table set schema
CREATE SCHEMA spcs2;
ALTER TABLE spcs1.a SET SCHEMA spcs2;
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert succeed
INSERT INTO a2 SELECT generate_series(1,200);
-- expect insert succeed
INSERT INTO spcs2.a SELECT generate_series(1,200);
ALTER TABLE spcs2.a SET SCHEMA spcs1;
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,200);
ERROR:  tablespace: schemaspc, schema: spcs1 diskquota exceeded
SELECT schema_name, tablespace_name, quota_in_mb, nspsize_tablespace_in_bytes FROM diskquota.show_fast_schema_tablespace_quota_view WHERE schema_name = 'spcs1' and tablespace_name ='schemaspc';
 schema_name | tablespace_name | quota_in_mb | nspsize_tablespace_in_bytes 
-------------+-----------------+-------------+-----------------------------
 spcs1       | schemaspc       |           1 |                     4030464
(1 row)

-- Test alter tablespace
-- start_ignore
\! mkdir -p /tmp/schemaspc2
-- end_ignore
DROP TABLESPACE  IF EXISTS schemaspc2;
NOTICE:  tablespace "schemaspc2" does not exist, skipping
CREATE TABLESPACE schemaspc2 LOCATION '/tmp/schemaspc2';
ALTER TABLE a SET TABLESPACE schemaspc2;
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert succeed
INSERT INTO a SELECT generate_series(1,200);
ALTER TABLE a SET TABLESPACE schemaspc;
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,200);
ERROR:  tablespace: schemaspc, schema: spcs1 diskquota exceeded
-- Test update quota config
SELECT diskquota.set_schema_tablespace_quota('spcs1', 'schemaspc', '10 MB');
 set_schema_tablespace_quota 
-----------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert success
INSERT INTO a SELECT generate_series(1,100);
-- expect insert success
INSERT INTO a SELECT generate_series(1,1000000);
SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert fail
INSERT INTO a SELECT generate_series(1,100);
ERROR:  tablespace: schemaspc, schema: spcs1 diskquota exceeded
-- Test delete quota config
SELECT diskquota.set_schema_tablespace_quota('spcs1', 'schemaspc', '-1 MB');
 set_schema_tablespace_quota 
-----------------------------
 
(1 row)

SELECT diskquota.wait_for_worker_new_epoch();
 wait_for_worker_new_epoch 
---------------------------
 t
(1 row)

-- expect insert success
INSERT INTO a SELECT generate_series(1,100);
-- start_ignore
\! mkdir -p /tmp/schemaspc3
-- end_ignore
DROP TABLESPACE  IF EXISTS "Schemaspc3";
NOTICE:  tablespace "Schemaspc3" does not exist, skipping
CREATE TABLESPACE "Schemaspc3" LOCATION '/tmp/schemaspc3';
CREATE SCHEMA "Spcs2";
SELECT diskquota.set_schema_tablespace_quota('"Spcs2"', '"Schemaspc3"', '-1 MB');
 set_schema_tablespace_quota 
-----------------------------
 
(1 row)

RESET search_path;
DROP TABLE spcs1.a2, spcs1.a;
DROP SCHEMA spcs1, spcs2;
DROP TABLESPACE schemaspc;
DROP TABLESPACE schemaspc2;
DROP TABLESPACE "Schemaspc3";
